portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -43,36 +46,73 @@
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int maxComments = Integer.parseInt(UtilityFunctions.getFieldFromResponse(request, "maxcomments", defaultMaxComment));
     String sortMetric = UtilityFunctions.getFieldFromResponse(request, "metric", "time");
     String sortOrder = UtilityFunctions.getFieldFromResponse(request, "order", "desc");
 
+    ArrayList<UserComment> comments = new ArrayList<>();
+    comments = populateRootComments(comments, maxComments, sortOrder, sortMetric);
+
+    Gson gson = new Gson();
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
+  }
+
+  /*
+   * Returns a list containing atmost maxComments top-level queries and all their replies. 
+   * The top-level queries are sorted by sortMetric in sortOrder
+   */ 
+  private ArrayList<UserComment> populateRootComments(ArrayList<UserComment> comments, int maxComments, String sortOrder, String sortMetric) {
     Datastore datastore = DatastoreOptions.getDefaultInstance().getService();
     Builder builder = Query.newEntityQueryBuilder();
-    builder = builder.setKind("Comment");
+    builder = builder.setKind("Comment").setFilter(PropertyFilter.eq("rootid", 0));
+
     if (sortOrder.equals("desc")) {
       builder = builder.setOrderBy(StructuredQuery.OrderBy.desc(sortMetric));
     } else {
       builder = builder.setOrderBy(StructuredQuery.OrderBy.asc(sortMetric));
     }
+    
+    builder = builder.setLimit(maxComments);
+
     Query<Entity> query = builder.build();
     QueryResults<Entity> results = datastore.run(query);
+    while (results.hasNext()) {
+      Entity entity = results.next();
+      UserComment comment = entityToComment(entity);
+      comments.add(comment);
+      comments = populateChildComments(comments, entity.getKey().getId());
```

Comment:
nit: you can make the populateChildComments return null because it is modifying the `comments` list already.